vllm/entrypoints/chat_utils.py

```diff
@@ -160,104 +236,77 @@ def _get_full_multimodal_text_prompt(placeholder_token_str: str,
 def _parse_chat_message_content_parts(
     role: str,
     parts: Iterable[ChatCompletionContentPartParam],
-    model_config: ModelConfig,
-    tokenizer: AnyTokenizer,
-) -> ChatMessageParseResult:
+    mm_tracker: MultiModalItemTracker,
+) -> List[ConversationMessage]:
     texts: List[str] = []
-    mm_futures: List[Awaitable[MultiModalDataDict]] = []
-    modality: Literal["image", "audio"] = "image"
+    mm_placeholders: List[str] = []
 
     for part in parts:
         part_type = part["type"]
         if part_type == "text":
             text = _TextParser.validate_python(part)["text"]
             texts.append(text)
         elif part_type == "image_url":
-            modality = "image"
-            if len(mm_futures) > 0:
-                raise NotImplementedError(
-                    "Multiple multimodal inputs is currently not supported.")
-
             image_url = _ImageParser.validate_python(part)["image_url"]
 
             if image_url.get("detail", "auto") != "auto":
                 logger.warning(
                     "'image_url.detail' is currently not supported and "
                     "will be ignored.")
 
-            image_future = async_get_and_parse_image(image_url["url"])
-            mm_futures.append(image_future)
+            image_coro = async_get_and_parse_image(image_url["url"])
+            placeholder = mm_tracker.add("image", image_coro)
+            if placeholder:
+                mm_placeholders.append(placeholder)
         elif part_type == "audio_url":
-            modality = "audio"
-            if len(mm_futures) > 0:
-                raise NotImplementedError(
-                    "Multiple multimodal inputs is currently not supported.")
-
             audio_url = _AudioParser.validate_python(part)["audio_url"]
-            audio_future = async_get_and_parse_audio(audio_url["url"])
-            mm_futures.append(audio_future)
+            audio_coro = async_get_and_parse_audio(audio_url["url"])
+            placeholder = mm_tracker.add("audio", audio_coro)
+            if placeholder:
+                mm_placeholders.append(placeholder)
         else:
             raise NotImplementedError(f"Unknown part type: {part_type}")
 
     text_prompt = "\n".join(texts)
+    if mm_placeholders:
+        text_prompt = _get_full_multimodal_text_prompt(mm_placeholders,
+                                                       text_prompt)
 
-    if mm_futures:
-        placeholder_token_str = _mm_token_str(model_config, tokenizer,
-                                              modality)
-        if placeholder_token_str is not None:
-            if placeholder_token_str in text_prompt:
-                logger.warning(
-                    "Detected multi-modal token string in the text prompt. "
-                    "Skipping prompt formatting.")
```

Comment:
IMO we should keep this logic. This is really useful in case the default placeholder padding (i.e, to the front of the prompt) doesn't work well with the model, which happens to some models in practice actually (chameleon, deepseek-vl, etc).

We could simply check if the first placeholder exists in the user prompt and assume the user intends to add the placeholder themselves if that's the case, and verify prompt contains what's in `mm_placeholders` as substrings. (perhaps we could use dictionary instead of list for `mm_placeholders`?)

Let me know if you have any strong opinion about this!

Reply:
That makes sense, but I think we should be able to do something simpler. I also realized we should apply this formatting after we apply chat template to the whole conversation instead of a part of the conversation since the multimodal data item limit is for the entire conversation.

Here are the pseudocode I'm thinking
```python
mm_placeholders: Dict[str, int] # key: placeholder_str, value: count
text_prompt: str # final prompt after we apply the chat template

# Check if the prompt has any of placeholder
missing_placeholders = []
for key in mm_placeholders:
    mm_placeholders[key] -= text_prompt.count(key)
   
    # Raise error if more placeholders are present in the prompt
    if mm_placeholders[key] < 0:
        raise

    # Add to missing placeholders list
    missing_placeholders += mm_placeholders[key] * [key]

final_prompt = "\n".join(missing_placeholders + [text_prompt])
```
How does this look?

Reply:
The one thing I wasn't sure about was a natural way to track the position within the former messages list once it's been stringified from the template -- unless I'm misunderstanding what you're proposing I think that would end up putting the images before the first message (e.g. system prompt) rather than at the start of the message in which they appeared, right?

(In any event, using `count()` is certainly simpler than what I wrote above, so whatever we decide I'll rewrite it to use that.)

Reply:
Ah yes you're right, and it looks like we're already passing `mm_tracker` across messages in a conversation so the count shouldn't be an issue! (I thought it was initialized for each message, sorry for not reading your code too carefully 😅 )

Reply:
Updated the pseudocode above!